### PR TITLE
fix for ICE issue #25180

### PR DESCRIPTION
**Pull request by llogiq**
According to @eddyb – and my tests – the following gets rid of the ICE in issue #25180.

### CODE REVIEW

**Comment by llogiq** [comment]
I think what the PR lacks as of yet is a testcase to show that the fix actually quells the ICE. I just added it and will commit after I rebuild.

File: src/librustc/middle/check_const.rs [hunk 1]
```diff
@@ -251,9 +251,13 @@ impl<'a, 'tcx, 'v> Visitor<'v> for CheckCrateVisitor<'a, 'tcx> {
                 b: &'v ast::Block,
                 s: Span,
                 fn_id: ast::NodeId) {
-        assert!(self.mode == Mode::Var);
+        let (old_mode, old_qualif) = (self.mode, self.qualif);
+        self.mode = Mode::Var;
+        self.qualif = ConstQualif::empty();
```
Comment (eddyb) [inline src/librustc/middle/check_const.rs]: You don't have to *copy* `with_mode`, you can just call it (and replace `self` in the two original statements with `v`, or whatever you choose for the name of the closure argument).

**Comment by llogiq** [comment]
Yep, you're right. I'll change it.

File: src/librustc/middle/check_const.rs [hunk 2]
```diff
@@ -251,8 +251,8 @@ impl<'a, 'tcx, 'v> Visitor<'v> for CheckCrateVisitor<'a, 'tcx> {
                 b: &'v ast::Block,
                 s: Span,
                 fn_id: ast::NodeId) {
-        assert!(self.mode == Mode::Var);
-        self.with_euv(Some(fn_id), |euv| euv.walk_fn(fd, b));
+        self.with_mode(Mode::Var, |v| v.with_euv(Some(fn_id),
+            |euv| euv.walk_fn(fd, b)));
         visit::walk_fn(self, fk, fd, b, s);
```
Comment (eddyb) [inline src/librustc/middle/check_const.rs]: `walk_fn` has to be inside the `with_mode` call, too.

**Comment by llogiq** [comment]
OK, but I'd like to understand why? Surely rustc compiles the code anyway – I even have a test case saying so. What are the implications of walking the fn in const mode?

**Comment by eddyb** [comment]
If it's in const mode, you get the const rules. And maybe ICEs. Try calling a function and you'll see.

**Comment by llogiq** [comment]
Ah, Ok. Should have run the extended test before I fixed the function.

Is this ready to merge now or do I need to do something else?

File: src/test/run-pass/issue-25180.rs [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+// pretty-expanded FIXME #25180
+
+const x: &'static Fn() = &|| println!("ICE here");
```
Comment (eddyb) [inline src/test/run-pass/issue-25180.rs]: Do multiple statements, arguments and local variables work in these closures?

**Comment by llogiq** [comment]
Yeah, it's a good idea to add some more tests. I'm on it.

**Comment by llogiq** [comment]
So yes, it [appears to work](https://github.com/llogiq/rust/blob/master/src/test/run-pass/issue-25180.rs).

**Comment by llogiq** [comment]
I guess I'll try and fix up the git history before you pull.

File: src/test/run-pass/issue-25180.rs [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+// pretty-expanded FIXME #25180
```
Comment (eddyb) [inline src/test/run-pass/issue-25180.rs]: The failure was in pretty tests... Which I don't you can do if you're using macros like `println!` which use unstable items in their expansions..

**Comment by eddyb** [comment]
@steveklabnik huh, shouldn't a test failure take it out of the queue automatically?

**Comment by steveklabnik** [comment]
don't think so

**Comment by llogiq** [comment]
I'm sorry that I broke the rollup. Apparently my tests weren't sufficient. I've now removed the print! macro invocations and added a feature declaration (which was hinted in the error message).

Can someone tell me how I can reproduce the failing test?

**Comment by steveklabnik** [comment]
It's all good :) everyone does it. I'm not 100% sure how to reproduce though

**Comment by llogiq** [comment]
Thanks for the kind words. My first PR and manage to hit all available roadblocks at once.

But hey, I learn a lot there days. :smile:

File: src/test/run-pass/issue-25180.rs [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+#![feature(std_misc)]
+
+const EMPTY: &'static Fn() = &|| ();
+
+const ONE_ARGUMENT: &'static Fn(u32) = &|y| assert!(y == 42);
```
Comment (eddyb) [inline src/test/run-pass/issue-25180.rs]: `assert!` still uses the formatting machinery. I'm curious if we prefer to disable pretty testing or just add feature gates. What do other tests do?

**Comment by llogiq** [comment]
I see 783 instances of "feature" just in `test/run-pass`. So yes, it appears feature-gates are not that uncommon. I'm unsure if #![feature(std_misc)] will be enough to make it work.

**Comment by eddyb** [comment]
@llogiq Running `make check-stage1-rpass-pretty` should answer that question.

**Comment by llogiq** [comment]
@eddyb Thanks, but that only prints that my Makefile does not have this target. Perhaps I'm missing some required package?

**Comment by pnkfelix** [comment]
@llogiq no I think @eddyb just mispoke.  The target is (probably) `make check-stage1-pretty-rpass`.

(I always recommend GNU remake for debugging big makefile systems.)

**Comment by eddyb** [comment]
@llogiq @pnkfelix My bad, I have both in my shell history and I pasted the bad one.

**Comment by llogiq** [comment]
In that case, my previous two commits make the test pass on my machine.

**Comment by llogiq** [comment]
I've just revisited the problem, and recent changes not only made my PR unmergeable, but also fixed the problem in a different way. Closing the PR is the sensible option here.

**Pull request closed without merging**